File: py/42.接雨水.py
#
# @lc app=leetcode.cn id=42 lang=python3
#
# [42] 接雨水
#

# @lc code=start
class Solution:
    #      * 暴力法 每个柱子的积水区，取决于左边界的最大值，右边界的最小值 两者中的较小者减去柱子本身的高度
    #  * 时间复杂度是n*n
    #  time limit exceeded
    def trap1(self, height: List[int]) -> int:
        res = 0
        for i in range(1,len(height)-1):
            left_max, right_max = 0, 0
            for j in range(i,-1,-1):
                left_max = max(left_max,height[j])
            for j in range(i,len(height)):
                right_max = max(right_max,height[j])
            res += (min(right_max,left_max)-height[i])
        return res
        
#  * 三趟扫描
#  * 第一次扫描每个柱子的左边界
#  * 第二次扫描每个柱子的右边界
#  * 第三次遍历数组，取左右边界高度的最小值减去柱子的高度
#  * 40ms   95%    7%  14MB
 
    def trap2(self, height: List[int]) -> int:
        if height is None or len(height)==0:
            return 0
        res = 0
        left_max_list = [-1] * len(height)
        right_max_list = [-1] * len(height)
        left_max_list[0] = height[0]
        right_max_list[-1] = height[-1]
        for i in range(1,len(height)):
            left_max_list[i] = max(left_max_list[i-1],height[i])
        for i in range(len(height)-2,-1,-1):
            right_max_list[i] = max(right_max_list[i+1],height[i])
        for i in range(len(height)):
            res += (min(left_max_list[i],right_max_list[i])-height[i])

        return res
    
#       * 用栈的存储左边界
#  * 56ms   37%   73%    13.7MB
    def trap3(self, height: List[int]) -> int:
        res = 0
        stack = []
        current = 0
        while current < len(height):
            while len(stack) >0  and height[current] > height[stack[-1]]:
                top = stack.pop()
                if len(stack)==0:
                    break
                distance = current - stack[-1] - 1
                res += (distance* (min(height[current],height[stack[-1]])-height[top]))
            stack.append(current)
            current += 1

        return res
#     * 双指针最快
#  * 44ms  97.99%   46%  13.8MB
    def trap(self, height: List[int]) -> int:
        res = 0
        left,right = 0,len(height) - 1
        left_max,right_max = 0,0
        while left < right:
            if height[left]<height[right]:
                if height[left] > left_max:
                    left_max = height[left]
                else:
                    res += left_max - height[left]
                left += 1
            else:
                if height[right] > right_max:
                    right_max = height[right]
                else:
                    res += right_max - height[right]
                right -= 1

        return res

# @lc code=end

File: java/42.接雨水.java
/*
 * @lc app=leetcode.cn id=42 lang=java
 *
 * [42] 接雨水
 */

// @lc code=start
class Solution {
    /**
     * 暴力法 每个柱子的积水区，取决于左边界的最大值，右边界的最小值 两者中的较小者减去柱子本身的高度
     * 时间复杂度是n*n
     * 111ms   5.34%    99.52%  37.5MB
     */
    public int trap1(int[] height) {
        int max = 0;
        for(int i=1;i<height.length-1;i++){
            int maxLeft=0,maxRight=0;
            for(int j=i;j<height.length;j++){
                maxRight = Math.max(height[j],maxRight);
            }
            for(int j=i;j>=0;j--){
                maxLeft = Math.max(height[j], maxLeft);
            }
            max += Math.min(maxLeft,maxRight) - height[i];

        }
        return max;

    }
/**
 * 三趟扫描
 * 第一次扫描每个柱子的左边界
 * 第二次扫描每个柱子的右边界
 * 第三次遍历数组，取左右边界高度的最小值减去柱子的高度
 * 2ms   45%    95%  37.9MB
 */
    public int trap2(int[] height) {
        if(height==null || height.length==0) return 0;
        int max = 0;
        int size = height.length;
        int[] leftMax = new int[size];
        int[] rightMax = new int[size];
        leftMax[0] = height[0];
        rightMax[size-1] = height[size-1];
        for(int i=1;i<size;i++){
            leftMax[i] = Math.max(height[i],leftMax[i-1]);
        }
        for(int i=size-2;i>=0;i--){
            rightMax[i] = Math.max(height[i],rightMax[i+1]);
        }
        for(int i=0;i<size;i++){
            max += Math.min(rightMax[i],leftMax[i]) - height[i];
        }
        
        return max;

    }
/**
 * 用栈的存储左边界
 * 2ms   45%   92%    38MB
 */
    public int trap3(int[] height) {
        if(height==null || height.length==0) return 0;
        int max = 0;
        Deque<Integer> stack = new LinkedList<Integer>();
        int curr = 0;
        while(curr<height.length){
            while(!stack.isEmpty()&&height[curr]>height[stack.peek()]){
                int top = stack.pop();
                if(stack.isEmpty())break;
                int distance = curr - stack.peek()-1;
                int boundedHeight = Math.min(height[curr],height[stack.peek()]) - height[top];
                max += boundedHeight*distance;
            }
            stack.push(curr++);
        }
        
        return max;
    }
/**
 * 双指针最快
 * 1ms  99.99%   88%  38.2MB
 */
    public int trap(int[] height) {
        int max = 0;
        int left =0,right=height.length-1;
        int leftMax=0,rightMax=0;
        while(left<right){
            if(height[left]<height[right]){
                if(height[left]>leftMax){
                    leftMax = height[left];
                }else{
                    max += (leftMax-height[left]);
                }
                left++;
            }else{
                if(height[right]>rightMax){
                    rightMax = height[right];
                }else{
                    max += (rightMax-height[right]);
                }
                right--;
            }
        }
        return max;
    }
}
// @lc code=end

